File: services/email-triage/triage/adapters/todoist.py
from __future__ import annotations

from dataclasses import dataclass

import httpx


@dataclass
class CreatedTask:
    id: str
    url: str


async def create_task(
    *,
    token: str,
    content: str,
    description: str | None = None,
    priority: int | None = None,
    due_string: str | None = None,
) -> CreatedTask:
    body: dict = {"content": content}
    if description:
        body["description"] = description
    if priority:
        body["priority"] = priority
    if due_string:
        body["due_string"] = due_string
    async with httpx.AsyncClient(timeout=15.0) as client:
        res = await client.post(
            "https://api.todoist.com/api/v1/tasks",
            headers={
                "authorization": f"Bearer {token}",
                "content-type": "application/json",
            },
            json=body,
        )
        res.raise_for_status()
        data = res.json()
    return CreatedTask(id=str(data["id"]), url=data.get("url", ""))

File: triage-worker/src/adapters/todoist.ts
import type { Env } from "../env";

export interface CreatedTask {
  id: string;
  url: string;
}

export async function createTask(
  env: Env,
  args: { content: string; description?: string; priority?: 1 | 2 | 3 | 4; dueString?: string },
): Promise<CreatedTask> {
  const res = await fetch("https://api.todoist.com/rest/v2/tasks", {
    method: "POST",
    headers: {
      authorization: `Bearer ${env.TODOIST_TOKEN}`,
      "content-type": "application/json",
    },
    body: JSON.stringify({
      content: args.content,
      description: args.description,
      priority: args.priority,
      due_string: args.dueString,
    }),
  });
  if (!res.ok) throw new Error(`todoist create failed: ${res.status} ${await res.text()}`);
  const data = (await res.json()) as { id: string; url: string };
  return { id: data.id, url: data.url };
}
